src/soc/amd/common/boot/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+#![feature(llvm_asm)]
+#![feature(lang_items, start)]
+#![no_std]
+#![no_main]
+#![feature(global_asm)]
+
+use arch::bzimage::BzImage;
+use arch::ioport::IOPort;
+use model::Driver;
+use uart::i8250::I8250;
+mod acpi;
+use acpi::setup_acpi_tables;
+extern crate heapless; // v0.4.x
+use heapless::consts::*;
+use heapless::Vec;
+
+use core::ptr;
+// Until we are done hacking on this, use our private copy.
+// Plan to copy it back later.
+global_asm!(include_str!("bootblock.S"));
+
+fn poke32(a: u32, v: u32) -> () {
+    let y = a as *mut u32;
+    unsafe {
+        ptr::write_volatile(y, v);
+    }
+}
+fn poke8(a: u32, v: u8) -> () {
+    let y = a as *mut u8;
+    unsafe {
+        ptr::write_volatile(y, v);
+    }
+}
+
+fn peek8(a: u32) -> u8 {
+    let y = a as *mut u8;
+    unsafe { ptr::read_volatile(y) }
+}
+
+/// Write 32 bits to port
+unsafe fn outl(port: u16, val: u32) {
+    llvm_asm!("outl %eax, %dx" :: "{dx}"(port), "{al}"(val));
+}
+
+/// Read 32 bits from port
+unsafe fn inl(port: u16) -> u32 {
+    let ret: u32;
+    llvm_asm!("inl %dx, %eax" : "={ax}"(ret) : "{dx}"(port) :: "volatile");
+    return ret;
+}
+fn peek32(a: u32) -> u32 {
+    let y = a as *const u32;
+    unsafe { ptr::read_volatile(y) }
+}
+// extern "C" {
+//     fn run32(w: &mut impl core::fmt::Write, start_address: usize, dtb: usize);
+// }
+
+fn peek(a: u64) -> u64 {
+    let y = a as *const u64;
+    unsafe { ptr::read_volatile(y) }
+}
+
+fn peekb(a: u64) -> u8 {
+    let y = a as *const u8;
+    unsafe { ptr::read_volatile(y) }
+}
+
+// Returns a slice of u32 for each sequence of hex chars in a.
+fn hex(a: &[u8], vals: &mut Vec<u64, U8>) -> () {
```

Comment:
This probably belongs in a library someday

Reply:
Probably--although maybe we can just use/pass the Writer anyway. Not that sureful to have

Anyway, for now, this patchset only moves stuff and doesn't change it if possible. That way, the move can be identified in the commit log later.